File: C++/binary tree from postorder.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left, *right;
    TreeNode(int val)
    {
        this->val = val;
        this->left = nullptr;
        this->right = nullptr;
    }

    TreeNode(int val, TreeNode *left, TreeNode *right) : val(val), left(left), right(right) {}
};

int postIndex;

TreeNode *helper(vector<int> &postorder, unordered_map<int, int> &inIndex, int start, int end)
{
    if (start > end)
        return nullptr;

    TreeNode *head = new TreeNode(postorder[postIndex--]);

    if (start == end)
        return head;

    int curr = inIndex[head->val];

    head->right = helper(postorder, inIndex, curr + 1, end);
    head->left = helper(postorder, inIndex, start, curr - 1);

    return head;
}

TreeNode *buildTree(vector<int> &postorder, vector<int> &inorder)
{
    unordered_map<int, int> inIndex;
    for (int i = 0; i < inorder.size(); i++)
        inIndex[inorder[i]] = i;
    postIndex = postorder.size() - 1;
    return helper(postorder, inIndex, 0, postorder.size() - 1);
}

void preorder(TreeNode *root)
{
    if (root == nullptr)
        return;

    cout << root->val << " ";
    preorder(root->left);
    preorder(root->right);
}

int main(int argc, char const *argv[])
{
    vector<int> inorder = {9, 3, 15, 20, 7};
    vector<int> postorder = {9, 15, 7, 20, 3};
    TreeNode *root = buildTree(postorder, inorder);
    preorder(root);
    cout << endl;
    return 0;
}

File: src/BinaryTreeFromPostorder.java
import java.util.Hashtable;

public class BinaryTreeFromPostorder {
    public static void main(String[] args) {
        int[] inorder = { 9, 3, 15, 20, 7 };
        int[] postorder = { 9, 15, 7, 20, 3 };
        TreeNode head = buildTree(postorder, inorder);
        TreeOperations.preorder(head);
        System.out.println();
    }

    public static TreeNode buildTree(int[] postorder, int[] inorder) {
        inIndex.clear();
        for (int i = 0; i < inorder.length; i++)
            inIndex.put(inorder[i], i);

        postIndex = postorder.length - 1;

        return helper(postorder, 0, inorder.length - 1);
    }

    private static int postIndex;
    private static Hashtable<Integer, Integer> inIndex = new Hashtable<>();

    private static TreeNode helper(int[] postorder, int start, int end) {
        if (start > end)
            return null;
        TreeNode head = new TreeNode(postorder[postIndex--]);

        if (start == end)
            return head;

        int curr = inIndex.get(head.val);

        head.right = helper(postorder, curr + 1, end);
        head.left = helper(postorder, start, curr - 1);

        return head;
    }
}
